servers/quarkus-authn/src/main/java/org/projectnessie/server/authn/NessieAuthMechanism.java

```diff
@@ -0,0 +1,132 @@
+import io.quarkus.vertx.http.runtime.security.ChallengeData;
+import io.quarkus.vertx.http.runtime.security.HttpAuthenticationMechanism;
+import io.quarkus.vertx.http.runtime.security.HttpCredentialTransport;
+import io.smallrye.mutiny.Uni;
+import io.vertx.ext.web.RoutingContext;
+import jakarta.enterprise.context.ApplicationScoped;
+import jakarta.inject.Inject;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.projectnessie.quarkus.config.QuarkusNessieAuthenticationConfig;
+
+/**
+ * HTTP authentication mechanism implementation to disallow all unauthorized accesses, except for
+ * the explicitly configured paths and static resources.
+ */
+@ApplicationScoped
+public class NessieAuthMechanism implements HttpAuthenticationMechanism {
+
+  /**
+   * {@code HttpAuthenticationMechanism}s with a higher priority are "asked" first, let
+   * `NessieAuthMechanism` be called quite late, basically as the last one.
+   */
+  public static final int PRIORITY = 0;
+
+  private final boolean enabled;
+  private final AuthenticationFailedException unauthorized;
+  private final Set<String> anonymousPaths;
+  private final Set<String> anonymousPathPrefixes;
+
+  @Inject
+  public NessieAuthMechanism(QuarkusNessieAuthenticationConfig config) {
+    System.err.println("NessieAuthEnabledMechanism: HELLO AUTH IS ENABLED");
+    this.enabled = config.enabled();
+    this.unauthorized = new AuthenticationFailedException("Missing or unrecognized credentials");
+    Set<String> paths = new HashSet<>();
+    config.anonymousPaths().ifPresent(paths::addAll);
+    config.anonymousPathPrefixes().ifPresent(paths::addAll);
+    this.anonymousPaths = paths;
+    this.anonymousPathPrefixes =
+        config
+            .anonymousPathPrefixes()
+            .map(prefixes -> prefixes.stream().map(p -> p + '/').collect(Collectors.toSet()))
+            .orElse(Set.of());
+  }
+
+  boolean checkPath(String path) {
+    // Allow certain preconfigured paths (e.g. health checks) to be serviced without
+    // authentication.
+    return anonymousPaths.contains(path)
+        || anonymousPathPrefixes.stream().anyMatch(path::startsWith);
+  }
+
+  @Override
+  public int getPriority() {
+    System.err.println("NessieAuthEnabledMechanism: getPriority");
+    return PRIORITY;
+  }
+
+  @Override
+  public Uni<SecurityIdentity> authenticate(
+      RoutingContext context, IdentityProviderManager identityProviderManager) {
+    System.err.println(
+        "NessieAuthEnabledMechanism: AUTHENTICATE "
+            + context.request().absoluteURI()
+            + " / "
+            + context.request().uri());
+
+    if (enabled) {
+      var path = context.request().path();
+      if (checkPath(path)) {
+        System.err.println("NessieAuthEnabledMechanism:   AUTHENTICATE ... ANONYMOUS REST");
+        return Uni.createFrom().nullItem();
+      }
+
+      if (!path.contains("..")) {
```

Comment:
Why do we need this?

Reply:
To prevent path traversals

Reply:
Let's make it in a dedicated PR for visibility / traceability... WDYT?

Reply:
We have to wait for Quarkus 3.16.2 with [quarkus#44279](https://github.com/quarkusio/quarkus/pull/44279) anyways...